### PR TITLE
universe: go: add git by default in go.#Image

### PR DESCRIPTION
**Pull request by vdemeester**
This adds git to the default `go.#Image` image so that 1.18 "just
works".

Fixes https://github.com/dagger/dagger/issues/1965

Signed-off-by: Vincent Demeester <vincent@sbr.pm>

### CODE REVIEW

**Comment by helderco** [comment]
I'd rather follow up on https://github.com/docker-library/golang/issues/415, and if you really need git, then you can:

```cue
img: go.#Image & {
    packages: git: {}
}
```

File: pkg/universe.dagger.io/go/image.cue
Comment (TomChv) [inline pkg/universe.dagger.io/go/image.cue:32]: Instead of adding whole `docker.#Run` operation, you could simply add `packages: "git": _` below `packages` declaration.

And it will be installed by the for loop below.

**Review by TomChv — CHANGES_REQUESTED** [review]
Thank you for the contribution, I've suggest a small improvement so avoid code duplication ;)

File: pkg/universe.dagger.io/go/image.cue
Comment (TomChv) [reply pkg/universe.dagger.io/go/image.cue:32]: But I agree with Helder, that would be better to rely on official go image and add any package that you need in a custom `go.#Image`

**Comment by vdemeester** [comment]
> I'd rather follow up on [docker-library/golang#415](https://github.com/docker-library/golang/issues/415), and if you really need git, then you can:
> 
> ```cue
> img: go.#Image & {
>     packages: git: {}
> }
> ```

So In general, I would say yes. The only "trouble" with that is that `go` 1.18 assumes git is there by default. See https://github.com/dagger/dagger/issues/1965 for the error. Maybe installing git should be done conditionally on 1.18 version only ?

File: pkg/universe.dagger.io/go/image.cue
Comment (vdemeester) [reply pkg/universe.dagger.io/go/image.cue:32]: I tend to agree with both of you (also, I learned the `: _` approach today yay).

**Comment by helderco** [comment]
There's other options, like setting `buildvcs=false` by default.

**Comment by helderco** [comment]
Also @vdemeester, if you're up for it, add a test which includes a `git init`'ed source dir to check if the build doesn't fail because of it.

**Comment by vdemeester** [comment]
@helderco I'll do this tomorrow 👼🏼 😉

**Comment by aluzzardi** [comment]
👍 on this. I just hit the problem myself, had to do a similar workaround.

Ideally the upstream go image would already include git, but if that doesn't come soon enough, we should go ahead with this

**Comment by shykes** [comment]
I think we should add the polyfill without waiting for upstream. We can always remove it later if it becomes redundant.

File: pkg/universe.dagger.io/go/test/build.cue
Comment (vdemeester) [inline pkg/universe.dagger.io/go/test/build.cue:20]: I probably should add a 2nd test entry here (one with, one without the flag), right ?

File: pkg/universe.dagger.io/go/image.cue
Comment (vdemeester) [inline pkg/universe.dagger.io/go/image.cue:15]: Added the FIXME 🙃

**Comment by helderco** [comment]
There's two levels to this:

1. https://github.com/docker-library/golang/issues/415
  **Will the upstream golang docker image include git?** Doesn't seem likely, they are waiting on the status of 👇 

2. https://github.com/golang/go/issues/51748
  **Will golang itself disable buildvcs by default in case `git` isn't in `$PATH`?** Very possible:
  
  > That sounds like a consensus to me.
  >
  > Option (3) it is! 👍
  
  _Originally posted by @bcmills in https://github.com/golang/go/issues/51748#issuecomment-1072724739_

---

So I'm partial to adding a shim (i.e., follow golang's lead), which means: don't add git, and set `-buildvcs=false` if it's not installed in the image.

1. If golang ends up defaulting to `-buildvcs=auto`, then it's easy to remove the shim;
2. If golang stays the course, then we either keep the convenience or get on board and remove the shim just the same.

**Comment by helderco** [comment]
There's an argument for making dagger have the convenience of adding `git` on top of the golang docker image, but:

**Pros:**

- Easiest fix, at least for now;
- More convinent for the most common use case;

**Cons**

- Harder to remove later if users come to depend on it, then we're the ones breaking stuff;
- You may be using mercurial or something else and we're back to square 1, but then you need to be aware of this change from the go language and either install your tool or add the flag;

**Comment by helderco** [comment]
Looking at your change I come to realize that our definition needs a better way to set flags. This can come in another PR.

From my previous comment, I say keep it simple for now:

1. Just add `git` to the image;
2. Don't set the flag, we'll allow for better user control soon;

File: pkg/universe.dagger.io/go/image.cue
Comment (helderco) [reply pkg/universe.dagger.io/go/image.cue:15]: Yes, keep this, but it's `Remove once golang image includes git (docker-library/golang#415) **or** go compiler is smart with -buildvcs (golang/go#51748)`.

**Comment by gerhard** [comment]
To fix [the failing Lint check](https://github.com/dagger/dagger/runs/5789751625?check_suite_focus=true#step:8:12), run `make cuelint` locally. This will fix the cue formatting:

```diff
$ git diff
diff --git a/pkg/universe.dagger.io/go/build.cue b/pkg/universe.dagger.io/go/build.cue
index 7e46d800..c16a687a 100644
--- a/pkg/universe.dagger.io/go/build.cue
+++ b/pkg/universe.dagger.io/go/build.cue
@@ -21,8 +21,8 @@ import (
        // Build tags to use for building
        tags: *"" | string

-    // Enable buildvcs
-    buildvcs: bool | *true
+       // Enable buildvcs
+       buildvcs: bool | *true

        // LDFLAGS to use for linking
        ldflags: *"" | string
@@ -39,12 +39,12 @@ import (
                command: {
                        args: [package]
                        flags: {
-                               build:      true
-                               "-v":       true
-                               "-tags":    tags
-                if buildvcs == false {
-                    "-buildvcs=false": true
-                }
+                               build:   true
+                               "-v":    true
+                               "-tags": tags
+                               if buildvcs == false {
+                                       "-buildvcs=false": true
+                               }
                                "-ldflags": ldflags
                                "-o":       "/output/"
                        }
diff --git a/pkg/universe.dagger.io/go/image.cue b/pkg/universe.dagger.io/go/image.cue
index 2a5ae101..d2f2f5bb 100644
--- a/pkg/universe.dagger.io/go/image.cue
+++ b/pkg/universe.dagger.io/go/image.cue
@@ -12,8 +12,8 @@ _#DefaultVersion: "1.18"
        version: *_#DefaultVersion | string

        packages: [pkgName=string]: version: string | *""
-    // FIXME Remove once golang image include 1.18 *or* go compiler is smart with -buildvcs
-    packages: "git": _
+       // FIXME Remove once golang image include 1.18 *or* go compiler is smart with -buildvcs
+       packages: git: _

        // FIXME Basically a copy of alpine.#Build with a different image
        // Should we create a special definition?
diff --git a/pkg/universe.dagger.io/go/test/build.cue b/pkg/universe.dagger.io/go/test/build.cue
index 4318e69b..785336df 100644
--- a/pkg/universe.dagger.io/go/test/build.cue
+++ b/pkg/universe.dagger.io/go/test/build.cue
@@ -16,8 +16,8 @@ dagger.#Plan & {

                simple: {
                        build: go.#Build & {
-                               source: client.filesystem."./data/hello".read.contents
-                buildvcs: false
+                               source:   client.filesystem."./data/hello".read.contents
+                               buildvcs: false
                        }

                        exec: docker.#Run & {
```

I use [vim-cue](https://github.com/jjo/vim-cue) to do this for me automatically, as I edit `cue` files.

I know that others are using [vscode-cue](https://github.com/cue-sh/vscode-cue) with success.

**Review by helderco — APPROVED** [review]

**Comment by helderco** [comment]
Ping @shykes

**Comment by vdemeester** [comment]
cc @aluzzardi

**Comment by shykes** [comment]
This seems like a useful short-term fix. We will not be able to remove `git` from the default image in this package, because as @helderco noted, that might break developers that depend on it. But I think it's fine to keep git installed forever. Remember that this is only for the *default image*. Developers can easily swap in their own image, with or without git.

**Review by shykes — APPROVED** [review]
lgtm

**Comment by shykes** [comment]
We're good to merge @helderco

**Comment by helderco** [comment]
It's blocked until @TomChv approves his requested changes.

**Review by aluzzardi — APPROVED** [review]

**Comment by aluzzardi** [comment]
@TomChv's changed were addressed -- merging

**Comment by helderco** [comment]
https://go-review.googlesource.com/c/go/+/398855/

**Comment by vdemeester** [comment]
@helderco will probably be fixed in 1.18.1 or something I guess ? 🤔

**Pull request merged**
